hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/NettyRpcConnection.java

```diff
@@ -339,14 +365,79 @@ public void operationComplete(ChannelFuture future) throws Exception {
     });
   }
 
+  private void writeAndFlushToChannel(Call call, Channel ch) {
+    if (ch == null) {
+      return;
+    }
+
+    scheduleTimeoutTask(call);
+    ch.writeAndFlush(call).addListener(new ChannelFutureListener() {
+      @Override
+      public void operationComplete(ChannelFuture future) {
+        // Fail the call if we failed to write it out. This usually because the channel is
+        // closed. This is needed because we may shutdown the channel inside event loop and
+        // there may still be some pending calls in the event loop queue after us.
+        if (!future.isSuccess()) {
+          call.setException(toIOE(future.cause()));
+        }
+      }
+    });
+  }
+
   @Override
   public void sendRequest(final Call call, HBaseRpcController hrc) {
-    execute(eventLoop, () -> {
-      try {
-        sendRequest0(call, hrc);
-      } catch (Exception e) {
-        call.setException(toIOE(e));
+    try {
+      sendRequest0(call, hrc);
+    } catch (Exception e) {
+      call.setException(toIOE(e));
+    }
+  }
+
+  /**
+   * HBaseClientPipelineFactory is the netty pipeline factory for this netty connection
+   * implementation.
+   */
+  private static class HBaseClientPipelineFactory extends ChannelInitializer<SocketChannel> {
+
+    private SSLContext sslContext = null;
+    private SSLEngine sslEngine = null;
+    private final String host;
+    private final int port;
+    private final Configuration conf;
+
+    public HBaseClientPipelineFactory(String host, int port, Configuration conf) {
+      this.host = host;
+      this.port = port;
+      this.conf = conf;
+    }
+
+    @Override
+    protected void initChannel(SocketChannel ch) throws X509Exception.SSLContextException {
+      ChannelPipeline pipeline = ch.pipeline();
+      if (conf.getBoolean(HBASE_CLIENT_NETTY_TLS_ENABLED, false)) {
+        initSSL(pipeline);
       }
-    });
+      pipeline.addLast("handler", new BufferCallBeforeInitHandler());
+    }
+
+    // The synchronized is to prevent the race on shared variable "sslEngine".
+    // Basically we only need to create it once.
+    private synchronized void initSSL(ChannelPipeline pipeline)
+      throws X509Exception.SSLContextException {
+      if (sslContext == null || sslEngine == null) {
```

Comment:
I think here we should share the SslContext instance. That means, we'd better create a SslContext instance per RpcClient, and pass it in when creating a NettyRpcConnection, and then just call its createSSLEngine method when initializing a Channel, where we do not need synchronized any more.